File: java-algo/src/main/java/com/leetcode/Problem907.java
package com.leetcode;

/**
 * 907. Sum of Subarray Minimums
 * https://leetcode.com/problems/sum-of-subarray-minimums/
 */
public class Problem907 {
    public int sumSubarrayMins(int[] arr) {
        int mod = (int) 1e9 + 7;
        long ans = 0L;
        for (int i = 0; i < arr.length; i++) {
            int left = i - 1;
            while (left >= 0 && arr[i] < arr[left]) {
                left--;
            }
            int right = i + 1;
            while (right < arr.length && arr[i] <= arr[right]) {
                right++;
            }
            ans += (long) arr[i] * (i - left) * (right - i);
            ans %= mod;
        }
        return (int) ans;
    }
}

File: python-algo/leetcode/problem_907.py
"""907. Sum of Subarray Minimums
https://leetcode.com/problems/sum-of-subarray-minimums/
"""
from typing import List


class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        base = int(1e9 + 7)
        stack = []
        ans = 0
        for r in range(len(arr) + 1):
            t = arr[r] if r < len(arr) else 0
            while stack and arr[stack[-1]] >= t:
                cur = stack.pop()
                l = stack[-1] if stack else -1
                ans = (ans + (cur - l) * (r - cur) * arr[cur]) % base
            stack.append(r)
        return ans
